[Wasm-GC] Fix exception validation when using GC

takikawa opened the pull request:
#### 34288ce1365e496e4fb45fdd8919917413d660f1
<pre>
[Wasm-GC] Fix exception validation when using GC
<a href="https://bugs.webkit.org/show_bug.cgi?id=265678">https://bugs.webkit.org/show_bug.cgi?id=265678</a>

Reviewed by Justin Michaud.

Adjusts checking of throw arguments and also LLInt catch generation to
work in the presence of GC types.

* JSTests/wasm/gc/exception.js: Added.
(module):
(testExceptionsWithGC):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isSubtype):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addCatchToUnreachable):

Canonical link: <a href="https://commits.webkit.org/271527@main">https://commits.webkit.org/271527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c914e1c648bd6dd340fb5982546922a8f7755a35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30907 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25838 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9119 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4394 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26097 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28649 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5786 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24414 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5062 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5163 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31593 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/24547 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31463 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27474 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5127 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29220 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6725 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/34998 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6861 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5582 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7570 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5646 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

takikawa commented on JSTests/wasm/gc/exception.js:
```diff
@@ -0,0 +1,38 @@
+  return new WebAssembly.Module(buffer);
+}
+
+// These are all binary tests because exception handling is now finalized and the GC reference intepreter does not support it.
```
"now finalized" -> "not finalized"

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3011,7 +3011,7 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
         for (unsigned i = 0; i < exceptionSignature.argumentCount(); ++i) {
             TypedExpression arg = m_expressionStack.at(m_expressionStack.size() - i - 1);
-            WASM_VALIDATOR_FAIL_IF(arg.type() != exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1), "The exception being thrown expects the argument at index ", i, " to be ", exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1), " but argument has type ", arg.type());
+            WASM_VALIDATOR_FAIL_IF(!isSubtype(arg.type(), exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1)), "The exception being thrown expects the argument at index ", i, " to be ", exceptionSignature.argumentType(exceptionSignature.argumentCount() - i - 1), " but argument has type ", arg.type());
```
This won't change the behavior of non-gc types, right?

takikawa replied:
It shouldn't, in that if neither GC nor typed funcrefs is turned on then `isSubtype` will eventually fallback to just type equality. (https://searchfox.org/wubkat/source/Source/JavaScriptCore/wasm/WasmFormat.h#338)

If we want to be careful, we could adjust `isSubtype` to explicitly check `Options::useWebAssemblyTypedFunctionReferences` and only do type equality if the flag is off.

justinmichaud reviewed (approved):
r=me

takikawa committed:
[Wasm-GC] Fix exception validation when using GC
https://bugs.webkit.org/show_bug.cgi?id=265678

Reviewed by Justin Michaud.

Adjusts checking of throw arguments and also LLInt catch generation to
work in the presence of GC types.

* JSTests/wasm/gc/exception.js: Added.
(module):
(testExceptionsWithGC):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isSubtype):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addCatchToUnreachable):

Canonical link: https://commits.webkit.org/271527@main
